File: modelman/src/modelman/providers/lifecycle/probe.py
"""HTTP polling primitives shared by provider lifecycle backends.

Port/model/warmup polling loops generalized from `bin/lib/*.sh` and
today's mtplx-only helpers in `lifecycle/__init__.py`. Nothing here is
provider-specific: callers pass in the URLs to poll.
"""

from __future__ import annotations

import json
import re
import subprocess
import time
import urllib.error
import urllib.request

from ...local_process import http_models_ids
from .envelope import LifecycleError

# bash: wait_for_port_closed url 5 = 5 x (1s curl + 0.2s sleep)
STOP_WAIT_TIMEOUT = 6.0
# bash: poll_until_down ... 0.5 120 (mlx_lm_server pre-bind wait)
PREBIND_WAIT_TIMEOUT = 60.0
# bash: ollama's 90 x (2s health + <=120s chat + 1s sleep), as one deadline
WARMUP_TIMEOUT = 300.0
# today's _wait_for_model default, unchanged
MODEL_LOAD_TIMEOUT = 300.0
# bash: wait_for url name 30 = 30 x (2s + 1s)
RESTORE_WAIT_TIMEOUT = 90.0


def wait_for_port_closed(url: str, timeout: float = 6.0) -> None:
    """Poll a localhost URL until it stops responding, or raise on timeout.

    A response — success or HTTP error status — means something is still
    listening. `HTTPError` is a `URLError`/`OSError` subclass, so it must
    be caught before the `URLError` catch below, or a server merely
    answering with a non-2xx status (e.g. 404 for a not-yet-ready path)
    would be misread as the port having closed. A read `TimeoutError` is
    also NOT a closed port: the listener accepted the connection and then
    stalled (hung, or draining under load) — exactly the "still holds the
    port" case this poll exists to catch. Only a connection-level failure
    (refused, reset, no route) means the port actually closed.
    """
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=1.0) as resp:  # noqa: S310 — localhost probe
                resp.read()
        except urllib.error.HTTPError:
            pass  # server answered (with an error status) — still open
        except urllib.error.URLError as exc:
            # urlopen wraps connection-level failures in URLError. A
            # connect timeout (reason is a TimeoutError) is ambiguous —
            # treat it as still open rather than risk spawning into a
            # held port.
            if isinstance(exc.reason, TimeoutError):
                pass
            else:
                return  # refused / reset / no route — port closed
        except TimeoutError:
            pass  # accepted the connection, then stalled mid-read — still open
        except ConnectionError:
            return  # reset at read time — the listener is dying or gone
        time.sleep(0.2)
    raise LifecycleError(f"port still answering at {url} after {timeout}s")


def port_closed_within(url: str, *, timeout: float) -> bool:
    """Same polling loop as `wait_for_port_closed`, but returns True/False
    instead of raising on timeout. Used by stop-and-wait helpers that want
    to warn rather than fail when a port doesn't close in time."""
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=1.0) as resp:  # noqa: S310 — localhost probe
                resp.read()
        except urllib.error.HTTPError:
            pass  # server answered (with an error status) — still open
        except urllib.error.URLError as exc:
            if isinstance(exc.reason, TimeoutError):
                pass
            else:
                return True  # refused / reset / no route — port closed
        except TimeoutError:
            pass  # accepted the connection, then stalled mid-read — still open
        except ConnectionError:
            return True  # reset at read time — the listener is dying or gone
        time.sleep(0.2)
    return False


def wait_for_port_open(url: str, *, timeout: float = 90.0, interval: float = 1.0) -> bool:
    """Poll a localhost URL until it answers (any response, including an
    HTTP error status, counts as "up"), returning True once it does or
    False after `timeout` elapses. Mirrors bash's `poll_until_up`, which
    never raises — this matches that: non-raising, boolean return."""
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=1.0) as resp:  # noqa: S310 — localhost probe
                resp.read()
            return True
        except urllib.error.HTTPError:
            return True  # server answered (with an error status) — up
        except OSError:
            pass  # not listening yet, or a transient connection failure
        time.sleep(interval)
    return False


def wait_for_model(
    models_url: str,
    model: str,
    *,
    proc: subprocess.Popen | None = None,
    timeout: float = 300.0,
) -> None:
    """Poll `models_url` until it lists `model`, the serve process dies
    (when `proc` is given), or the deadline passes."""
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if proc is not None and proc.poll() is not None:
            # The serve process died mid-load (OOM kill, missing weights
            # discovered late): without this check the HTTP poll below
            # runs the full timeout against a dead port while the real
            # crash cause stays buried in the log.
            raise LifecycleError(f"serve process exited during model load (exit {proc.returncode})")
        ids = http_models_ids(models_url)
        if any(served == model or served.endswith("/" + model) for served in ids):
            return
        time.sleep(1.0)
    raise LifecycleError(f"timed out waiting for {models_url} to serve {model}")


def serving_model(models_url: str, model: str) -> bool:
    """True when `models_url` already lists `model` (same lenient match as
    `wait_for_model`). `http_models_ids` returns [] on any error, so a down
    or wedged server reads as "not serving" — the safe fallback."""
    ids = http_models_ids(models_url)
    return any(served == model or served.endswith("/" + model) for served in ids)


def warmup(chat_url: str, model: str, *, health_url: str, timeout: float = 300.0) -> None:
    """Warm a model into GPU/RAM: poll `health_url` for liveness (matching
    bash's `curl -m 2` probe), then POST a 1-token chat completion to
    `chat_url` and check for the chat.completion marker in the body.

    Supports asymmetric health/chat URLs (e.g. ollama's health check hits
    `/api/tags` while its chat completions go to `/v1/chat/completions`) —
    today's mtplx-only `_warmup` conflated the two because mtplx's health
    check was baked in elsewhere.
    """
    payload = json.dumps(
        {
            "model": model,
            "messages": [{"role": "user", "content": "hi"}],
            "max_tokens": 1,
            "temperature": 0,
            "stream": False,
        }
    ).encode()
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            urllib.request.urlopen(health_url, timeout=2.0)  # noqa: S310 — localhost probe
        except OSError:
            time.sleep(1.0)
            continue
        try:
            req = urllib.request.Request(
                chat_url,
                data=payload,
                headers={"Content-Type": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=120) as resp:  # noqa: S310
                body = resp.read().decode()
            # MTPLX (and others) may serialize with spaces between keys and
            # values (e.g. `"object": "chat.completion"`); tolerate
            # arbitrary whitespace around the colon rather than requiring
            # compact JSON.
            if re.search(r'"object"\s*:\s*"chat\.completion"', body):
                return
        except OSError:
            pass
        time.sleep(1.0)
    raise LifecycleError(f"failed to warm up model {model} at {chat_url}")

File: wt/internal/lifecycle/probe.go
package lifecycle

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/ohanaverse/local-ai-setup/wt/internal/config"
	"github.com/ohanaverse/local-ai-setup/wt/internal/localmodels"
)

// procWatch lets waitForModel notice that a serve process died mid-load.
type procWatch interface {
	exited() (done bool, err error)
}

var chatCompletionMarker = regexp.MustCompile(`"object"\s*:\s*"chat\.completion"`)

// probe does one GET. responded is true for ANY HTTP response (2xx or an error
// status: something is listening); timedOut is true when the listener accepted
// the connection and then stalled. A refused/reset connection is neither.
func (e *env) probe(ctx context.Context, url string, timeout time.Duration) (responded, timedOut bool) {
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return false, false
	}
	resp, err := e.probeClient.Do(req)
	if err != nil {
		var ne net.Error
		if errors.As(err, &ne) && ne.Timeout() || errors.Is(err, context.DeadlineExceeded) {
			return false, true
		}
		return false, false
	}
	_ = resp.Body.Close()
	return true, false
}

func (e *env) sleep(ctx context.Context, d time.Duration) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(d):
		return nil
	}
}

// waitPortOpen polls url until it answers (any response counts), returning
// false (no error) when timeout passes first; ctx cancellation returns its error.
func (e *env) waitPortOpen(ctx context.Context, url string, timeout time.Duration) (bool, error) {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if err := ctx.Err(); err != nil {
			return false, err
		}
		if responded, _ := e.probe(ctx, url, time.Second); responded {
			return true, nil
		}
		if err := e.sleep(ctx, e.pollInterval); err != nil {
			return false, err
		}
	}
	return false, ctx.Err()
}

// portClosedWithin polls url until it stops answering. A response — success or
// error status — or a read timeout means something still holds the port; only
// a connection-level failure means it closed.
func (e *env) portClosedWithin(ctx context.Context, url string, timeout time.Duration) (bool, error) {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if err := ctx.Err(); err != nil {
			return false, err
		}
		responded, timedOut := e.probe(ctx, url, time.Second)
		if err := ctx.Err(); err != nil {
			return false, err
		}
		if !responded && !timedOut {
			return true, nil
		}
		if err := e.sleep(ctx, e.pollInterval); err != nil {
			return false, err
		}
	}
	return false, ctx.Err()
}

// waitForModel polls modelsURL until it lists model (lenient name match), the
// serve process dies (when proc is given), or timeout passes.
func (e *env) waitForModel(ctx context.Context, modelsURL, model string, proc procWatch, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if err := ctx.Err(); err != nil {
			return err
		}
		if proc != nil {
			if done, err := proc.exited(); done {
				return fmt.Errorf("serve process exited during model load: %v", err)
			}
		}
		for _, id := range localmodels.FetchModelIDs(e.probeClient, modelsURL) {
			if localmodels.NameMatches(id, model) {
				return nil
			}
		}
		if err := e.sleep(ctx, e.pollInterval); err != nil {
			return err
		}
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	return fmt.Errorf("timed out waiting for %s to serve %s", modelsURL, model)
}

// warmup forces model into memory: poll healthURL for liveness, then POST a
// 1-token chat completion to chatURL and require the chat.completion marker.
func (e *env) warmup(ctx context.Context, chatURL, model, healthURL string, timeout time.Duration) error {
	payload, _ := json.Marshal(map[string]any{
		"model":       model,
		"messages":    []map[string]string{{"role": "user", "content": "hi"}},
		"max_tokens":  1,
		"temperature": 0,
		"stream":      false,
	})
	deadline := time.Now().Add(timeout)
	var lastReason string
	for time.Now().Before(deadline) {
		if err := ctx.Err(); err != nil {
			return err
		}
		if responded, _ := e.probe(ctx, healthURL, 2*time.Second); !responded {
			lastReason = "health check at " + healthURL + " did not respond"
			if err := e.sleep(ctx, e.pollInterval); err != nil {
				return err
			}
			continue
		}
		ok, reason := e.tryChat(ctx, chatURL, payload)
		if ok {
			return nil
		}
		lastReason = reason
		if err := e.sleep(ctx, e.pollInterval); err != nil {
			return err
		}
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	if lastReason == "" {
		lastReason = "no attempts completed"
	}
	return fmt.Errorf("failed to warm up model %s at %s: %s", model, chatURL, lastReason)
}

func (e *env) tryChat(ctx context.Context, chatURL string, payload []byte) (ok bool, reason string) {
	ctx, cancel := context.WithTimeout(ctx, 120*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, chatURL, bytes.NewReader(payload))
	if err != nil {
		return false, err.Error()
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := e.chatClient.Do(req)
	if err != nil {
		return false, err.Error()
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, "reading response: " + err.Error()
	}
	// A non-2xx answer is not a warmup, even when its body echoes a
	// chat.completion marker: the ported probe raises HTTPError here and
	// retries. Accepting it reports a model as resident that is not.
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return false, fmt.Sprintf("HTTP %d: %s", resp.StatusCode, truncateForError(body))
	}
	if !chatCompletionMarker.Match(body) {
		return false, "response missing chat.completion marker: " + truncateForError(body)
	}
	return true, ""
}

// truncateForError bounds a server response body embedded in an error
// message: a warmup failure's body can be an HTML error page or a large
// JSON blob, and the caller's final error must stay a readable one-liner.
func truncateForError(b []byte) string {
	const max = 300
	// strings.Fields+Join below is O(len(b)); warmup polls once a second for
	// up to warmupTimeout (600s), so a failing server that keeps answering
	// with a multi-MB body (a misrouted proxy, an unrelated HTTP service on
	// the probed port) would otherwise pay a full-body allocate-and-split
	// pass on every poll purely to produce a 300-character message.
	// Whitespace collapsing only removes characters, never adds them, so
	// bounding the input to a fixed multiple of max bytes before splitting
	// still leaves far more than enough to fill max runes.
	const scanLimit = max * 8
	if len(b) > scanLimit {
		b = []byte(trimTrailingPartialRune(string(b[:scanLimit])))
	}
	// strings.Fields splits on any whitespace (including newlines/tabs) and
	// drops empty runs; joining with a single space collapses the body to
	// one line and trims its ends, so an HTML error page or a formatted
	// stack trace can't carry its line breaks into the caller's one-line
	// error message.
	s := strings.Join(strings.Fields(string(b)), " ")
	if len(s) <= max {
		return s
	}
	return trimTrailingPartialRune(s[:max]) + "…"
}

// trimTrailingPartialRune backs off to a valid rune boundary: DecodeLastRuneInString
// reports (RuneError, 1) only when the trailing byte(s) are not a complete,
// valid encoding — the impossible-for-correct-UTF-8 signal that a byte cut
// landed mid-rune.
func trimTrailingPartialRune(s string) string {
	for len(s) > 0 {
		r, size := utf8.DecodeLastRuneInString(s)
		if r != utf8.RuneError || size != 1 {
			break
		}
		s = s[:len(s)-1]
	}
	return s
}

// liveServed asks a single-model provider's server directly what it is serving,
// and reports whether that answer can be trusted.
//
// known is false only when the server failed to give a usable answer — it
// accepted the connection and stalled, or answered with a non-2xx or
// undecodable body. A refused or reset connection is NOT unknown: it means
// nothing is listening, so known is true with no ids, and an ordinary cold
// start proceeds without a confirmation. Collapsing those two cases either
// refuses every cold start or permits the silent replacement this exists to
// prevent.
func (e *env) liveServed(ctx context.Context, cfg *config.Config, family string) (ids []string, known bool) {
	origin, _ := localmodels.FamilyOrigin(cfg, family)
	modelsURL := origin + "/v1/models"
	responded, timedOut := e.probe(ctx, modelsURL, e.prebindTimeout)
	switch {
	case responded:
		ids, err := localmodels.FetchModelIDsErr(e.probeClient, modelsURL)
		if err != nil {
			return nil, false
		}
		return ids, true
	case timedOut:
		return nil, false
	default:
		// Connection refused/reset: definitively nothing listening.
		return nil, true
	}
}
